allennlp/modules/transformer/bert_self_attention.py

```diff
@@ -0,0 +1,64 @@
+import math
+
+import torch
+
+from allennlp.common import FromParams
+
+
+class BertSelfAttention(torch.nn.Module, FromParams):
+    def __init__(
+        self, hidden_size: int, num_attention_heads: int, dropout: float = 0.0,
+    ):
+        super().__init__()
+        if hidden_size % num_attention_heads != 0:
+            raise ValueError(
+                "The hidden size (%d) is not a multiple of the number of attention "
+                "heads (%d)" % (hidden_size, num_attention_heads)
+            )
+        self.num_attention_heads = num_attention_heads
+        self.attention_head_size = int(hidden_size / num_attention_heads)
+        self.all_head_size = self.num_attention_heads * self.attention_head_size
+
+        self.query = torch.nn.Linear(hidden_size, self.all_head_size)
+        self.key = torch.nn.Linear(hidden_size, self.all_head_size)
+        self.value = torch.nn.Linear(hidden_size, self.all_head_size)
+
+        self.dropout = torch.nn.Dropout(dropout)
+
+    def transpose_for_scores(self, x):
+        new_x_shape = x.size()[:-1] + (self.num_attention_heads, self.attention_head_size,)
+        x = x.view(*new_x_shape)
+        return x.permute(0, 2, 1, 3)
```

Comment:
I guess `BiAttention` could derive from `SelfAttention` so that we can share this function. But maybe that's too much machinery to save four lines of code.